tests/components/mysensors/test_init.py

```diff
@@ -0,0 +1,252 @@
+"""Test MySensors base component."""
+import asyncio
+import sys
+from unittest.mock import MagicMock, call, patch
+
+import pytest
+import voluptuous as vol
+
+from homeassistant.components import mysensors as mysensors_comp
+from homeassistant.setup import async_setup_component
+
+from tests.common import mock_coro_func
+from tests.components.mysensors import (
+    DEVICE, MockChild, MockGateway, MockMQTTGateway, MockNode, get_gateway,
+    setup_mysensors)
+
+# pylint: disable=redefined-outer-name
+
+
+@pytest.fixture
+def gateway(loop, hass, mock_isdevice, mock_mysensors, mock_gw_ready):
+    """Set up mysensors component and return a gateway instance."""
+    res = loop.run_until_complete(setup_mysensors(hass))
+    assert res
+    loop.run_until_complete(hass.async_block_till_done())
+    gateway = get_gateway(hass)
+    return gateway
+
+
+@pytest.fixture
+def mock_isdevice():
+    """Mock isdevice."""
+    with patch('homeassistant.helpers.config_validation.isdevice',
+               return_value=DEVICE) as mock_test:
+        yield mock_test
+
+
+@pytest.fixture
+def mock_mysensors():
+    """Mock mysensors library."""
+    sys.modules['mysensors'] = MagicMock()
+    sys.modules['mysensors.mysensors'] = MagicMock()
+    import mysensors.mysensors as mysensors
+    mysensors.AsyncMQTTGateway = MockMQTTGateway
+    mysensors.AsyncSerialGateway = MockGateway
+    mysensors.AsyncTCPGateway = MockGateway
+
+    return mysensors
+
+
+@pytest.fixture
+def mock_gw_ready(loop):
+    """Mock gateway ready future."""
+    fut = asyncio.Future(loop=loop)
+    fut.set_result(True)
+    with patch('homeassistant.components.mysensors.gateway.asyncio.Future',
+               return_value=fut) as mock_future:
+        yield mock_future
+
+
+async def test_no_mqtt_config(hass, mock_mysensors):
+    """Test setup of mysensors with mqtt device but without mqtt config."""
+    config = {'mysensors': {
+        'gateways': [{'device': 'mqtt'}], 'version': '2.0',
+        'persistence': False}}
+
+    with patch('homeassistant.components.mysensors.gateway'
+               '.async_setup_component',
+               side_effect=mock_coro_func()) as mock_stp_comp:
+        mock_stp_comp.return_value = False
+        res = await async_setup_component(hass, 'mysensors', config)
+    assert not res
+
+
+async def test_bad_device(hass, mock_mysensors):
+    """Test setup of mysensors with bad device referens."""
+    config = {'mysensors': {
+        'gateways': [{'device': ''}], 'version': '2.0', 'persistence': False}}
```

Comment:
Limiting the lines does not improve readability.

Reply:
What do you mean?

Reply:
That I rather see each key on a new line.